Add slice_ functions

wvictor14 opened the pull request:
Hi I found this [issue](https://github.com/tidyomics/genomics-todos/issues/6) on tidyomics and thought it would be good experience to try

Let me know how it looks and if there's any changes needed

This adds the slice_ family:

- slice_min
- slice_max
- slice_head
- slice_tail

stemangiola commented on R/dplyr_methods.R:
```diff
@@ -911,40 +911,144 @@ slice_sample.Seurat <- function(.data, ..., n = NULL, prop = NULL, by = NULL, we
+
+#' @export
+slice_head.Seurat <- function(.data, ..., n, prop, by = NULL) {
+    new_meta <- dplyr::slice_head(.data[[]], ..., n = n, prop = prop, by = {{by}})
```
To save memory for huge dataset, could we 
- somehow select 0 columns from this dataset .data[[]] ? Would the dataset still be valid?
- instead of new_meta we could just get rownames ?

What if the metadata does not have row names? For other tidy adapters we have used the trick of rowid_to_column(name = "row_number___"), then used this index to subset Seurat object .data[...,]

This would apply to all new functions.

wvictor14 replied:
What do you think of this,

a. https://github.com/stemangiola/tidyseurat/pull/68/commits/1d2ed45aa10733367da2d6733ebfe0d709557630 extract entire metadata first but then subset to needed columns, for `slice` the only column needed would be the optional argument `{{ by }}`. Other slice functions like _min _max could have more variables

and use `rowid_to_column` in case there are no rownames

another approach:

b. https://github.com/stemangiola/tidyseurat/pull/68/commits/d0b85918f01f1d0411dfb21878118ca80b95f747 construct an empty data.frame with no columns using rownames of .data

add in necessary variables for slice, so `{{ by }}` and arrange_by variables

perform slice operation and return row index / row names


b. is more memory efficient than a., due to never extracting the entire metadata df. But the implementation is a bit more verbose and complex

stemangiola replied:
I would say `a` is good 

- select columns
- `rowid_to_column` in any case
- `slice_*`

Great!

stemangiola commented on R/dplyr_methods.R:
```diff
@@ -867,36 +867,35 @@ full_join.Seurat <- function (x, y, by = NULL, copy = FALSE, suffix = c(".x", ".
-  #new_obj@meta.data = new_meta
-
+  idx <- .data[[]] |>
+    tibble::rowid_to_column(var  = 'row_number___')  |>
+    dplyr::select(-everything(), row_number___, {{ .by }}) |>
```
Thanks a lot!

Sorry if I am being pedantic here, but it would be better to do `select` first and `rowid_to_column` later both from a memory and elegance perspective.

Another little feedback,  standard we use is to use 

`@importFrom tibble ...`

and not use 

`tibble:: ...`

With these two changes I think the PR is good to go!

wvictor14 replied:
No problem, will do

wvictor14 commented on R/dplyr_methods.R:
```diff
@@ -917,29 +935,27 @@ slice_sample.Seurat <- function(.data, ..., n = NULL, prop = NULL, by = NULL, we
-    new_obj@meta.data =
-      new_meta %>%
-      data.frame(row.names=pull(.,!!c_(.data)$symbol), check.names = FALSE) %>%
-      select(- !!c_(.data)$symbol)
```
I am not sure why adding the metadata is necessary after subset?
@stemangiola

stemangiola commented:
@wvictor14 have a look at the github action; it is failing. Check maybe on your local CHECK and BiocCheck because here I have to "allow" github action each time, unfortunately. After you become a contributor, github action will be runnable by you in the future.

stemangiola commented:
@wvictor14  please add your authorship details here https://docs.google.com/spreadsheets/d/19XqhN3xAMekCJ-esAolzoWT6fttruSEermjIsrOFcoo/edit?usp=sharing

wvictor14 commented:
@stemangiola I fixed the errors for CRAN check. There are two bioccheck errors that I am not sure best to address:

* ERROR: Maintainer must add package name to Watched Tags on the support
      site; Edit your Support Site User Profile to add Watched Tags.
      
 and 

ERROR: At least 80% of man pages documenting exported objects must have
      runnable examples.

I realized that the option of inputting a tibble to `order_by` argument for `slice_min` and `slice_max` doesn't work nicely with the `select |> slice` approach. So I added an internal function `return_args` that returns the variables as symbols of such cases https://github.com/stemangiola/tidyseurat/pull/68/commits/7cca3f8116b4ce137c807db953d5de19f0a3d412

stemangiola commented:
> There are two bioccheck errors that I am not sure best to address

No problem, that could be another "challenge"



> Also I realized that the option of inputting a tibble to `order_by` argument for `slice_min` and `slice_max` doesn't work with the `select |> slice` approach

Could you please give me minimal example ?

wvictor14 commented:
> > Also I realized that the option of inputting a tibble to `order_by` argument for `slice_min` and `slice_max` doesn't work with the `select |> slice` approach
> 
> Could you please give me minimal example ?

This is what I mean

```
# in action
pbmc_small |> slice_min(tibble::tibble(nFeature_RNA, nCount_RNA), n = 2)

# return_args extracts the variables out of an expression
order_by <- expr(tibble::tibble(nFeature_RNA, nCount_RNA))
order_by_vars <- return_args(!!order_by)

# let's you input into select
pbmc_small |> select(!!!order_by_vars) |>

  # and then can apply slice_min
  slice_min(!!order_by, n = 6)

# A tibble: 6 × 2
  nFeature_RNA nCount_RNA
         <int>      <dbl>
1           26         51
2           29        157
3           29        172
4           30        150
5           30        202
6           31         62
```

Not sure if there is a more straightforward way, or how to make it more robust. What do you think?

stemangiola commented:
> > > Also I realized that the option of inputting a tibble to `order_by` argument for `slice_min` and `slice_max` doesn't work with the `select |> slice` approach
> > 
> > 
> > Could you please give me minimal example ?
> 
> This is what I mean
> 
> ```
> # in action
> pbmc_small |> slice_min(tibble::tibble(nFeature_RNA, nCount_RNA), n = 2)
> 
> ```

I still don't see in what scenario we would like to do this programmatically

> ```
> # return_args extracts the variables out of an expression
> order_by <- expr(tibble::tibble(nFeature_RNA, nCount_RNA))
> order_by_vars <- return_args(!!order_by)
> 
> # let's you input into select
> pbmc_small |> select(!!!order_by_vars) |>
> 
>   # and then can apply slice_min
>   slice_min(!!order_by, n = 6)
> ```

Rather than 

```r
pbmc_small |> 
   select(nFeature_RNA, nCount_RNA) |>
 
 # and then can apply slice_min
 slice_min(tibble::tibble(nFeature_RNA, nCount_RNA), n = 6)
```

stemangiola commented:
> > > Also I realized that the option of inputting a tibble to `order_by` argument for `slice_min` and `slice_max` doesn't work with the `select |> slice` approach
> > 
> > 
> > Could you please give me minimal example ?
> 
> This is what I mean
> 
> ```
> # in action
> pbmc_small |> slice_min(tibble::tibble(nFeature_RNA, nCount_RNA), n = 2)
> 
> # return_args extracts the variables out of an expression
> order_by <- expr(tibble::tibble(nFeature_RNA, nCount_RNA))
> order_by_vars <- return_args(!!order_by)
> 
> # let's you input into select
> pbmc_small |> select(!!!order_by_vars) |>
> 
>   # and then can apply slice_min
>   slice_min(!!order_by, n = 6)
> 
> # A tibble: 6 × 2
>   nFeature_RNA nCount_RNA
>          <int>      <dbl>
> 1           26         51
> 2           29        157
> 3           29        172
> 4           30        150
> 5           30        202
> 6           31         62
> ```
> 
> Not sure if there is a more straightforward way, or how to make it more robust. What do you think?

Great! I left a last comment to address (Above) and then this PR is on its way!

wvictor14 commented:
>  A detail. In the tidy transcriptomics framework, we adopt the standard of, no abbreviations and no acronyms.
> 
> Would you mind changing return_args into a self-explanatory function name, e.g. return_arguments_of_... or anything else that would make sense.
> 
> This also applies to any variable, e.g. args_expr, args_vars, ...

I think I got them all!

stemangiola commented:
Thanks a lot @wvictor14  for your PR! Well done. 

We are using a "no-export" strategy to our packages, e.g.:

https://github.com/stemangiola/tidySingleCellExperiment/blob/master/NAMESPACE#L41

In case you were interested, as you already have experience in `tidyseurat` I just point to the cool challenge about adapting `tidyseurat` documentation to `tidySingleCellExperiment`, which will solve your and other exports or generics, to avoid conflicts when `tidyomics` package is loaded.

https://github.com/stemangiola/tidyseurat/issues/66

Of course no expectations.
